tests/test-sdi-progress-window.c

```diff
@@ -0,0 +1,318 @@
+     "A window with a progress bar must appear with the name and icon of a "
+     "snapped application.",
+     "Click on the 'Hide' button. The window must "
+     "disappear and no new window should appear (check below this one). "
+     "Confirm that is correct.",
+     test_manual_hide},
+    {5, "/progress_window/test_dual_progress_bar_1",
+     "A window with a progress bar must appear with the name and icon of a "
+     "snapped application, growing from 1 to 10 and going back to 1 "
+     "periodically. It must not 'pulse' (move left-to-right and then "
+     "right-to-left softly).\n"
+     "The text in the progress bar should be 'Description for task X (X/10)', "
+     "being X a number between 1 and 10.\nIf no window is shown, check if it "
+     "is below this window.",
+     "Confirm that is correct.", test_dual_progress_bar1},
+    {6, "/progress_window/test_dual_progress_bar_2",
+     "A second progress bar must appear in the same window with the name and "
+     "icon of a "
+     "snapped application, growing from 1 to 10 and going back to 1 "
+     "periodically. It must not 'pulse' (move left-to-right and then "
+     "right-to-left softly).\n"
+     "The text in the progress bar should be 'Description for task X (X/10)', "
+     "being X a number between 1 and 10.\nIf no window is shown, check if it "
+     "is below this window.",
+     "Confirm that is correct.", test_dual_progress_bar2},
+    {7, "/progress_window/test_dual_progress_bar_3",
+     "The first progress bar must disappear, remaining only the second "
+     "progress bar. Neither a new window, nor a new progress bar, should "
+     "appear (check below "
+     "this window just in case).\n",
+     "Confirm that is correct.", test_dual_progress_bar3},
+    {8, "/progress_window/test_dual_progress_bar_4",
+     "The window must disappear and no new window should appear (check below "
+     "this window just in case).\n",
+     "Confirm that is correct.", test_dual_progress_bar4},
+    {-1, NULL, NULL, NULL, NULL}};
+
+/**
+ * GApplication callbacks
+ */
+
+static void do_startup(GObject *object, gpointer data) {
+  progress_window = sdi_progress_window_new(G_APPLICATION(object));
+
+  // this gsettings key stores the path to the .desktop files of the programs
+  // in the dock. Since we need to test a progress bar in the dock, we need
+  // that list.
+  g_autoptr(GSettings) settings = g_settings_new("org.gnome.shell");
+  GList *applist = g_app_info_get_all();
+  g_autoptr(GStrvBuilder) builder = g_strv_builder_new();
+  // make a list with only snapped icons
+  for (GList *p = applist; p != NULL; p = p->next) {
+    GAppInfo *app_info = G_APP_INFO(p->data);
+    if (!g_app_info_should_show(app_info)) {
+      continue;
+    }
+    const gchar *command_line = g_app_info_get_commandline(app_info);
+    if (strstr(command_line, "/snapd/") != NULL) {
+      g_strv_builder_add(builder, g_app_info_get_id(p->data));
+    }
+  }
```

Comment:
mnh, shouldn't we override in tests only the content of this key so that it surely contains a such desktop file?

It also should be a file that is provided by the test suite no?

Reply:
The problem is that we need to use .desktop files from snaps to ensure that everything works as expected. That's why the test is taking all the installed programs.

Anyway, remember that these tests are meant to be done "manually".

Reply:
Or am I misunderstanding what you mean?

Reply:
Well, I would like to see if the tests could be automated in a place where no snap is installed too, and so having local test  files that are equal to all the potential snap files makes it more reproducible and allows to check all the cases we care about and adjust the ones we're managing wrongly in case there's one.

Reply:
The window and dock classes should be quite "stable", so I'm not worried for these tests to be only "manually runable". The complex part is in `sdi-refresh-monitor`, and the idea with this and the previous refactors is, precisely, to fully isolate that code from Gtk and the desktop and be able to do automatic tests in that part.

Reply:
In fact, the only way of running both the window, dock and notifications tests in an automated fashion would be using OpenQA, because they require to check things on windows and notifications, and to do clicks over them.

Maybe the notifications test could be automated without OpenQA by emulating all the DBus notifications API, but I think that it is too much work for a piece of code that is designed to not require changes.

Reply:
Mmm... maybe I can play with the environment variables... I'll do a test.

Reply:
@3v1n0 Ok, done. If you can review the change when you have some spare time, please.